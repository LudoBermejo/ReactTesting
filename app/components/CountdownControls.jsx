import React from 'react';

export default class CountdownControls extends React.Component {

  constructor(props) {
    super(props);
    this.renderStartStopButton = this.renderStartStopButton.bind(this);
  }

  renderStartStopButton(countdownStatus) {
    if (countdownStatus === 'started') {
      return (
        <button
          id="btnPause"
          className="button secondary"
          onClick={this.props.onClickPause()}
        >Pause</button>
      );
    } else if (countdownStatus === 'paused') {
      return (
        <button
          id="btnResume"
          className="button primary"
          onClick={this.props.onClickResume()}
        >Resume</button>
      );
    }
    return '';
  }

  render() {
    const { countdownStatus } = this.props;
    return (
      <div className="controls">
        { this.renderStartStopButton(countdownStatus) }
        <button
          id="btnClear"
          className="button alert"
          onClick={this.props.onClickStop()}
        >Clear</button>
      </div>
    );
  }
}

CountdownControls.propTypes = {
  countdownStatus: React.PropTypes.string.isRequired,
  onClickPause: React.PropTypes.func.isRequired,
  onClickResume: React.PropTypes.func.isRequired,
  onClickStop: React.PropTypes.func.isRequired
};
