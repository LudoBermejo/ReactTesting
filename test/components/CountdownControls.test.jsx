import React from 'react';
import renderer from 'react-test-renderer';
import { shallow, mount } from 'enzyme';
import CountdownControls from 'CountdownControls';

describe('Countdown form component', () => {
  it('renders correctly', () => {
    const tree = renderer.create(<CountdownControls countdownStatus="started"
                                                    onClickPause={() => {}}
                                                    onClickResume={() => {}}
                                                    onClickStop={() => {}} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  describe('Render buttons', () => {
    it('Button show pause if countdownStatus is started', () => {
      const countDownControls = shallow(<CountdownControls countdownStatus="started"
                                                      onClickPause={() => {}}
                                                      onClickResume={() => {}}
                                                      onClickStop={() => {}} />);

      const buttons = countDownControls.find('button');
      expect(buttons.nodes[0].props.id).toBe('btnPause');
    });

    it('Button show resume if countdownStatus is paused', () => {
      const countDownControls = shallow(<CountdownControls countdownStatus="paused"
                                                           onClickPause={() => {}}
                                                           onClickResume={() => {}}
                                                           onClickStop={() => {}} />);

      const buttons = countDownControls.find('button');
      expect(buttons.nodes[0].props.id).toBe('btnResume');
    });

    it('Only cancel button if not paused or resumed or started', () => {
      const countDownControls = shallow(<CountdownControls countdownStatus="random"
                                                           onClickPause={() => {}}
                                                           onClickResume={() => {}}
                                                           onClickStop={() => {}} />);

      const buttons = countDownControls.find('button');
      expect(buttons.nodes.length).toBe(1);
    });
  })

  describe('Click buttons', () => {
    it('Launch onClickPause if countdownStatus is started', () => {
      const onClickPause = jest.fn();
      const countDownControls = shallow(<CountdownControls countdownStatus="started"
                                                           onClickPause={onClickPause}
                                                           onClickResume={() => {
                                                           }}
                                                           onClickStop={() => {
                                                           }}/>);

      const buttons = countDownControls.find('button').first();
      buttons.simulate('onClick');

      expect(onClickPause).toHaveBeenCalled();
    });

    it('Launch onClickResume if countdownStatus is paused', () => {
      const onClickResume = jest.fn();
      const countDownControls = shallow(<CountdownControls countdownStatus="paused"
                                                           onClickResume={onClickResume}
                                                           onClickPause={() => {
                                                           }}
                                                           onClickStop={() => {
                                                           }}/>);

      const buttons = countDownControls.find('button').first();
      buttons.simulate('onClick');

      expect(onClickResume).toHaveBeenCalled();
    });

    it('Launch onClickStop when click on button stop', () => {
      const onClickStop = jest.fn();
      const countDownControls = shallow(<CountdownControls countdownStatus="random"
                                                           onClickStop={onClickStop}
                                                           onClickPause={() => {
                                                           }}
                                                           onClickResume={() => {
                                                           }}/>);

      const buttons = countDownControls.find('button').first();
      buttons.simulate('onClick');

      expect(onClickStop).toHaveBeenCalled();
    });
  })


});
