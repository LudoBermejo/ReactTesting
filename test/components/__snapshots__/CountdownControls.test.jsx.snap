exports[`Countdown form component renders correctly 1`] = `
<div
  className="controls">
  <button
    className="button secondary"
    id="btnPause"
    onClick={undefined}>
    Pause
  </button>
  <button
    className="button alert"
    id="btnClear"
    onClick={undefined}>
    Clear
  </button>
</div>
`;
